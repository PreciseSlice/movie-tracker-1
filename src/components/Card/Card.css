.card {
  flex-shrink: 0;
  border-radius: 3px;
  margin: 20px 50px;
  width: 260px;
  background-size: cover;
  height: 400px; }

.movie-title {
  height: 65px;
  font-size: 1.7em;
  width: 314px;
  margin: 0 auto; }

.favorite-btn {
  background-image: url("../../assets/heart-outline.svg");
  background-repeat: no-repeat;
  border: none;
  width: 30px;
  height: 30px; }
  .favorite-btn:hover {
    background-image: url("../../assets/heart-filled.svg");
    width: 30px;
    height: 30px;
    background-size: cover; }

.favorite {
  background-image: url("../../assets/heart-filled.svg");
  width: 30px;
  height: 30px;
  background-size: cover; }

#not-signed-in {
  display: flex;
  justify-content: center;
  align-items: center;
  color: #fff;
  border-radius: 3px;
  background-color: #9c7e92;
  width: 260px;
  height: 43px;
  position: absolute;
  bottom: 0px;
  margin: 0px 0 0 50px;
  z-index: 1; }
