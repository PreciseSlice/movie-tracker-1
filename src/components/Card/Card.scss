.card {
  flex-shrink: 0;
  border-radius: 3px;
  margin: 20px 50px;
  width: 260px;
  background-size: cover;
  height: 400px;
}

.movie-title {
  height: 65px;
}

.favorite {
  background-image: url('../../assets/heart-outline.svg');
  background-repeat: no-repeat;
  border: none;
  width: 30px;
  height: 30px;
  &:hover {
    background-image: url('../../assets/heart-filled.svg');
    width: 30px;
    height: 30px;
    background-size: cover;
  }
}