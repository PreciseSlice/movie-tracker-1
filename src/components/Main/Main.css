.main {
  display: flex;
  padding: 15px 0;
  overflow-x: scroll; }

.main div {
  margin: 12px 0 0 0; }

.view-favorites {
  position: absolute;
  right: 20px;
  height: 25px;
  border: none;
  background-color: #ceb3c5;
  color: #fff;
  font-size: 0.9em;
  border-radius: 3px; }
  .view-favorites:hover {
    background-color: #9c7e92; }
