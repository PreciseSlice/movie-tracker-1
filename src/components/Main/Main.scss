.main {
  display: flex;
  padding: 15px 0;
  overflow-x: scroll;
}

.main div {
  margin: 32px 0 0 0;
  position: relative;
}

.view-favorites {
  position: absolute;
  right: 20px;
  height: 25px;
  border: none;
  background-color: #ceb3c5;
  color: #fff;
  font-size: 0.9em;
  border-radius: 3px;
  &:hover {
    background-color: #9c7e92;
  }
}