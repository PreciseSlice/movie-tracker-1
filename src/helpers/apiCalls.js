fetchApi = async (url) => {
    try {
        const initialFetch = await fetch(url);
        if(initialFetch.status <= 200) {
            return await initialFetch.json();
        } else {
            throw new Error('Bad staus code!')
        }
    } catch (error) {
        throw new Error(`Error retrieving Movies: ${error}`);
    }
}

export default fetchApi;