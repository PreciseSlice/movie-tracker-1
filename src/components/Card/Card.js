import React, { Component } from 'react';
import './Card.css';
import PropTypes from 'prop-types';

let message;

class Card extends Component {
  constructor(props) {
    super(props);
    this.state = {
      clicked: false
    };
  }

  displaySignUp = loggedIn => {
    const { toggleFavorite, movieData } = this.props;
    loggedIn && toggleFavorite(movieData);
    message = !loggedIn ? <div id="not-signed-in">Sign up or log in to add favorites</div> : <div />;
    !loggedIn && this.setState({ clicked: true });
  };

  render() {
    const { movieData, loggedIn } = this.props;
    // eslint-disable-next-line
    const { title, overview, poster_path } = movieData;
    const favClass = movieData.favorite ? 'favorite' : '';
    return (
      <div>
        <h1 className="movie-title">{title}</h1>
        <article
          className="card"
          style={{ backgroundImage: `url(${poster_path})` }}
        />
        {!loggedIn && message}
        <button
          onClick={() => this.displaySignUp(loggedIn)}
          className={`favorite-btn ${favClass}`}
        />
      </div>
    );
  }
}

Card.propTypes = {
  movieData: PropTypes.object.isRequired,
  toggleFavorite: PropTypes.func,
  loggedIn: PropTypes.bool,
};

export default Card;
