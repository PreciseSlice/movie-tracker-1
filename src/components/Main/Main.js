import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
  getMoviesFromApi,
  setFavorites,
  addUser,
  getMovies
} from '../../actions/index.js';
import './Main.css';
import PropTypes from 'prop-types';
import {
  fetchApi,
  postBackend,
  deleteFromBackend
} from '../../helpers/apiCalls';
import Card from '../Card/Card';

export class Main extends Component {
  componentDidMount = async () => {
    this.props.fetchMovies();
  };

  postFavorite = async movieData => {
    const { userId: user_id } = this.props;
    const favoriteProperties = {
      movie_id: movieData.movie_id,
      user_id,
      title: movieData.title,
      poster_path: movieData.poster_path,
      release_date: movieData.release_date,
      vote_average: movieData.vote_average,
      overview: movieData.vote_average
    };
    await postBackend('api/users/favorites/new/', favoriteProperties);
  };

  toggleFavorite = async movieData => {
    const { userId, setFavorites } = this.props;
    const backendFavorites = await fetchApi(`api/users/${userId}/favorites/`);
    const existingFavorites = this.markFavsAsFavorites(backendFavorites.data);
    if (!movieData.favorite) {
      movieData.favorite = !movieData.favorite;
      this.postFavorite(movieData);
      setFavorites([...existingFavorites, movieData]);
    } else {
      const body = { id: userId, movie_id: movieData.movie_id };
      movieData.favorite = !movieData.favorite;
      this.deleteFromStore(existingFavorites, movieData);
      deleteFromBackend(
        `api/users/${userId}/favorites/${movieData.movie_id}`,
        body
      );
    }
  };

  markFavsAsFavorites = favorites => {
    return favorites.map(fav => {
      fav.favorite = true;
      return fav;
    });
  };

  deleteFromStore = (favorites, duplicate) => {
    const { setFavorites } = this.props;
    const duplicateRemoved = favorites.filter(fav => {
      return fav.movie_id !== duplicate.movie_id;
    });
    setFavorites(duplicateRemoved);
  };

  render() {
    let { movieData, loggedIn } = this.props;
    movieData = movieData ? movieData : [];
    if (movieData.length) {
      const movies = movieData.map(movie => {
        return (
          <Card
            movieData={movie}
            key={movie.movie_id}
            loggedIn={loggedIn}
            toggleFavorite={this.toggleFavorite}
          />
        );
      });
      return (
        <div className="main">
          {loggedIn && (
            <Link to={{ pathname: '/favorites' }} >
              <button className="view-favorites">
                Favorites
              </button>
            </Link>
          )}
          {movies}
        </div>
      );
    } else {
      return null;
    }
  }
}

export const mapStateToProps = state => ({
  loggedIn: state.loggedIn,
  userId: state.userData.userId
});

export const mapDispatchToProps = dispatch => ({
  fetchMovies: movieData => dispatch(getMoviesFromApi(movieData)),
  setFavorites: favorites => dispatch(setFavorites(favorites)),
  addUser: user => dispatch(addUser(user))
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);

Main.propTypes = {
  movieData: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      movie_id: PropTypes.number.isRequired,
      poster_path: PropTypes.string.isRequired
    })
  ),

  fetchMovies: PropTypes.func.isRequired,
  loggedIn: PropTypes.bool.isRequired,
  setFavorites: PropTypes.func.isRequired
};
