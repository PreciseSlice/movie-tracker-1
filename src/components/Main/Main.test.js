/* eslint-disable */
import React, { Component } from 'react';
import Card from '../Card/Card';
import { getMoviesFromApi, setFavorites } from '../../actions/index.js';
import { Main, mapStateToProps, mapDispatchToProps } from './Main';
import apiCalls from '../../helpers/apiCalls';
import { shallow } from 'enzyme';

// jest.mock('../../helpers/apiCalls');

describe('Main', () => {
  const fakeStore = (state) => {
    return {
      dispatch: () => {}
    };
  }
  let mockFn;
  let wrapper;
  let mockData;

  beforeEach(() => {
    mockFn = jest.fn();
    window.fetch = jest.fn().mockImplementation( (url) => {
      return Promise.resolve({
        status: 200,
        json: () => Promise.resolve(mockData)
      })
    })
    wrapper = shallow(<Main fetchMovies={mockFn}
                            setFavorites={mockFn}
                            loggedIn={false} />);
  })

  it('matches the snapshot', () => {
    
    expect(wrapper).toMatchSnapshot();
  }); 

  it('calls fetchMovies when component mounts', () => {
    expect(wrapper.instance().props.fetchMovies).toHaveBeenCalled()
  })

  describe('mapStateToProps', () => {
    it('returns an object with keys of loggedIn and userId', () => {
      const mockState = {
        loggedIn: false,
        userData: {
          userId: '3'
        }
      }
      expect(mapStateToProps(mockState)).toEqual({loggedIn: false, userId: '3'})
    })
  })

  describe('mapDispatchToProps', () => {
    it.skip('returns an object with keys of fetchMovies and setFavorites', () => {
      const mockDispatch = jest.fn().mockImplementation((action) => {
        return action;
      });

      expect(mapDispatchToProps(mockDispatch)).toEqual({fetchMovies: 'function', setFavorites: 'function'})
    })
  })

  describe('toggleFavorite', () => {
      mockData = {
        data: [
          {
            id: '3',
            movie_id: '5'
          }
        ]
      }
    it('calls fetch', async () => {
      expect(window.fetch).not.toHaveBeenCalled();
      await wrapper.instance().toggleFavorite({id: '0'});
      expect(window.fetch).toHaveBeenCalled();
    })

    it('calls postFavorites if selected card is not a duplicate', async () => {
      expect(wrapper.instance().props.postFavorites).not.toHaveBeenCalled;
      await wrapper.instance().toggleFavorite({id: '0'});
      expect(wrapper.instance().props.postFavorites).toHaveBeenCalled;
    })

    it('doesn\'t call postFavorites if selected card is a duplicate', async () => {
      expect(wrapper.instance().props.postFavorites).not.toHaveBeenCalled;
      await wrapper.instance().toggleFavorite({id: '3'});
      expect(wrapper.instance().props.postFavorites).not.toHaveBeenCalled;
    })
  })
});

