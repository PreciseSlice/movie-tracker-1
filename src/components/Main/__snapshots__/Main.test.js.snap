// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main matches the snapshot 1`] = `""`;
